activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -147,9 +147,18 @@ def first!
     def last(limit = nil)
       return find_last(limit) if loaded?
 
-      result = order_values.empty? && primary_key ? order(arel_table[primary_key].desc) : reverse_order
-      result = result.limit!(limit || 1)
-      limit ? result.reverse : result.first
+      initial_limit = limit_value
+      sql_limit = [limit || 1, initial_limit || Float::INFINITY].min
+
+      result = limit(sql_limit)
+      if initial_limit
+        result.offset!((offset_value || 0) + initial_limit - sql_limit)
```

Comment:
In this patch implementation people have a choice between `relation.last` and `relation.to_a.last`. In your variant, there will be no such choice. I vote for current version, but open for an argument.